jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/AsyncContent.java

```diff
@@ -53,22 +53,13 @@ public class AsyncContent implements Content.Sink, Content.Source, Closeable
      * <ul>
      * <li>immediately with a failure when this instance is closed or already in error</li>
      * <li>successfully when the {@link Content.Chunk} returned by {@link #read()} is released</li>
-     * <li>successfully just before the {@link Content.Chunk} is returned if the latter {@link Content.Chunk#hasRemaining() has no remaining byte}</li>
+     * <li>successfully just before the {@link Content.Chunk} is returned by {@link #read()},
+     * if the chunk {@link Content.Chunk#canRetain() cannot be retained}</li>
      * </ul>
      */
     @Override
     public void write(boolean last, ByteBuffer byteBuffer, Callback callback)
     {
-        // Since the contract is that the callback has to be succeeded when
-        // the chunk returned by read() is released, and since it is not
-        // possible to create chunks with no remaining byte, when the byte
-        // buffer is empty we need to replace it with EOF / EMPTY and cannot
-        // be notified about the release of the latter two.
-        // This is why read() succeeds the callback if it has no remaining
-        // byte, meaning it is either EOF or EMPTY. The callback is succeeded
-        // once and only once, but that happens either during read() if the
-        // byte buffer is empty or during Chunk.release() if it contains at
-        // least one byte.
         Content.Chunk chunk;
         if (byteBuffer.hasRemaining())
```

Comment:
Isn't `Content.Chunk.from` already doing that `byteBuffer.hasRemaining()` check?